libexec/rtld-elf/aarch64/reloc.c

```diff
@@ -167,6 +169,228 @@ _rtld_relocate_nonplt_self(Elf_Dyn *dynp, Elf_Auxinfo *aux)
+	struct tramp trampolines[];	/* Points to start of trampolines */
+};
+
+SLIST_HEAD(tramp_pgs, tramp_pg);
+
+#define DEFAULT_STACK_TABLE_SIZE 2
+
+typedef uintptr_t *tramp_stk_table_t;
+
+static void
+get_rstk(uint32_t index, tramp_stk_table_t table, const void *target)
+{
+	size_t len = cheri_getlen(table) / sizeof(*table);
+
+	if (index < len) {
+
+		size_t size;
+		char *stk;
+		struct Struct_Stack_Entry *entry;
+		Obj_Entry *dst;
+
+		size = 0x40000 * getpagesize();
+		stk = mmap(NULL,
+			   size,
+			   PROT_READ | PROT_WRITE,
+			   MAP_ANON | MAP_PRIVATE | MAP_STACK,
+			   -1, 0);
+		if (stk == MAP_FAILED)
+			rtld_die();
+		stk = cheri_clearperm(stk, CHERI_PERM_EXECUTIVE | CHERI_PERM_SW_VMEM) + size;
+		struct {
+			uint8_t generation;
+			uintptr_t top;
+		} *metadata = (void *)stk;
+		metadata[-1].generation = 0;
+		metadata[-1].top = (uintptr_t)&metadata[-1];
+
+		table[index] = (uintptr_t)stk;
+
+		entry = xmalloc(sizeof(*entry));
+		entry->stack = stk;
+
+		dst = obj_from_addr(target);
+		lockinfo.wlock_acquire(dst->stackslock);
+		SLIST_INSERT_HEAD(&dst->stacks, entry, link);
+		lockinfo.lock_release(dst->stackslock);
+
+	} else {
+
+		size_t new_len;
+		tramp_stk_table_t new_t;
+
+		new_len = len * 2;
+		new_t = xcalloc(new_len, sizeof(*new_t));
+		if (!new_t)
+			rtld_die();
+		assert(cheri_getlen(new_t) / sizeof(*new_t) == new_len);
+
+		for (size_t i = 0; i < len; ++i)
+			new_t[i] = table[i];
+
+		asm ("msr	ctpidr_el0, %0" :: "C" (new_t));
+		free(table);
+	}
+}
+
+static void (*thr_thread_start)(struct pthread *);
+
+void
+_rtld_thread_start(struct pthread *curthread)
+{
+	tramp_stk_table_t tls;
+	asm ("mrs	%0, ctpidr_el0" : "=C" (tls));
+	asm ("msr	rctpidr_el0, %0" :: "C" (tls));
+
+	tls = xcalloc(DEFAULT_STACK_TABLE_SIZE, sizeof(*tls));
+	tls[0] = (uintptr_t)get_rstk;
+	asm ("msr	ctpidr_el0, %0" :: "C" (tls));
+
+	thr_thread_start(curthread);
+}
+
+void
+_rtld_thread_start_init(void (*p)(struct pthread *))
+{
+	assert(!(cheri_getperm(p) & CHERI_PERM_EXECUTIVE));
+	assert(!thr_thread_start);
+	thr_thread_start = tramp_pgs_append(p, obj_from_addr(p));
+}
+
+static void (*thr_sighandler)(int, siginfo_t *, void *);
+
+void
+_rtld_sighandler(int sig, siginfo_t *info, void *_ucp)
+{
+	uintptr_t csp, rcsp;
+	ucontext_t *ucp = _ucp;
+
+	csp = ucp->uc_mcontext.mc_capregs.cap_sp;
+	asm ("mrs	%0, rcsp_el0" : "=C" (rcsp));
+	ucp->uc_mcontext.mc_capregs.cap_sp = rcsp;
+
+	thr_sighandler(sig, info, ucp);
+
+	rcsp = ucp->uc_mcontext.mc_capregs.cap_sp;
+	asm ("msr	rcsp_el0, %0" :: "C" (rcsp));
+	ucp->uc_mcontext.mc_capregs.cap_sp = csp;
+}
+
+void
+_rtld_sighandler_init(void *p)
+{
+	assert(!(cheri_getperm(p) & CHERI_PERM_EXECUTIVE));
+	assert(!thr_sighandler);
+	thr_sighandler = tramp_pgs_append(p, obj_from_addr(p));
+}
+
+static void *
+partition(struct tramp **inout, size_t len)
+{
+	void *hi = *inout;
+	void *lo = cheri_setbounds(hi, len);
+	ptraddr_t top, alignment;
+	size_t rem;
+
+	*inout = lo;
```

Comment:
This would be clearer if you did the lo = cheri_setbounds here, it's easy to miss that by putting it as an initialiser...